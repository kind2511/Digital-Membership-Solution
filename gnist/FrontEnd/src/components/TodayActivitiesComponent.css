.today-activities-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding-top: 400px; 
  
}
  
.today-activities-list {
  width: 100%;
  max-width: 600px;
  display: flex;
  flex-direction: column;
  align-items: center; 
  background-color: #F4F4F4;
  padding-top: 30px;
  border-radius: 20px;
  padding-right: 22px;
  padding-left: 22px;
  border:None;
}

  
.today-activity-entry {
  background-color: #FFFFFF;
  border: 1px solid #D9E3F0;
  padding: 16px;
  margin-bottom: 16px;
  border-radius: 80px;
  display: flex;
  align-items: center;
  cursor: pointer;
  width: 90%; 
  color: #32325D;
}
  
  .today-activity-image {
    width: 60px;
    height: 60px;
    border-radius: 50%;
    object-fit: cover;
    margin-right: 20px;
  }
  
  .today-activity-info {
    display: flex;
    flex-direction: column;
  }
  
  .today-activity-title {
    font-weight: bold;
  }
  
  .today-activity-text {
    font-size: 0.8rem;
  }
  
  .today-details-modal {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.5);
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 2;
  }
  
  .today-modal-content {
    background: #FFFFFF;
    padding: 20px;
    border-radius: 20px;
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
    width: 80%;
    max-width: 500px;
    max-height: 80vh;
    overflow-y: auto;
    color: #32325D;
    border: 1px solid #D9E3F0;
    position: relative;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
  }
  
  .today-modal-buttons {
    display: flex;
    justify-content: space-between;
    margin-top: 20px;
  }
  
  .today-modal-button {
    padding: 10px 20px;
    cursor: pointer;
    background-color: RGB(19, 182, 253);
    color: #FFFFFF;
    border: none;
    border-radius: 5px;
    margin-right: 10px;
  }
  
  .today-modal-button:hover {
    background: #004080;
    color: #FFFFFF;
  }
  
  .promo-sentence-container {
    height: 50px; 
    display: flex;
    justify-content: center;
    align-items: center;
    overflow: hidden; /* Prevents content from spilling out */
    color:#32325D;
  }
  
  .promo-sentence {
    font-size: 24px;
    font-weight: bold;
    color: black; 
    margin-bottom: 20px; 
    opacity: 0;
    transition: opacity 5s ease-in-out;
    font-family:fantasy;    
  }
  
  .promo-sentence.visible {
    opacity: 2;
  }
  
  .message.confirmation {
    background-color: red; 
}

.message.error {
    background-color: red; 
}
