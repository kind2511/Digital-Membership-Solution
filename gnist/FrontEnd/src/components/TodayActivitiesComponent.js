import React, { useState, useEffect, useCallback } from 'react';
import axios from 'axios';
import { useAuth0 } from '@auth0/auth0-react';
import './TodayActivitiesComponent.css';

const promoSentences = ["Dette skjer idag.."];

function TodayActivitiesComponent() {
    const { user, isAuthenticated } = useAuth0();
    const [activities, setActivities] = useState([]);
    const [selectedActivity, setSelectedActivity] = useState(null);
    const [displayedText, setDisplayedText] = useState('');
    const [sentenceIndex, setSentenceIndex] = useState(0);
    const [todayDate, setTodayDate] = useState(null);
    const [message, setMessage] = useState('');
    const [messageType, setMessageType] = useState('');
    const baseApiUrl = 'http://127.0.0.1:8000';

    const fetchTodayActivities = useCallback(async () => {
        try {
            const response = await fetch(`${baseApiUrl}/digital_medlemsordning/get_activity_today/`);
            const data = await response.json();
            const updatedActivities = data.map(activity => ({
                ...activity,
                isUserSignedUp: activity.signed_up_members.some(member => member.auth0ID === user.sub)
            }));
            setActivities(updatedActivities);
            setTodayDate(data.length > 0 ? data[0].date : new Date().toISOString().split('T')[0]);
        } catch (error) {
            console.error("Failed to fetch today's activities:", error);
            setActivities([]);
        }
    }, [user.sub]);

    useEffect(() => {
        fetchTodayActivities();
    }, [fetchTodayActivities]);

    useEffect(() => {
        const timerId = setInterval(() => {
            setDisplayedText(promoSentences[sentenceIndex]);
            setSentenceIndex((sentenceIndex + 1) % promoSentences.length);
        }, 3000);
        return () => clearInterval(timerId);
    }, [sentenceIndex]);

    const handleActivityClick = (activity) => {
        setSelectedActivity(activity);
    };

    const handleSignUp = async () => {
        if (!isAuthenticated) {
            setMessage('Du må være logget inn for å melde på.');
            setMessageType('error');
            setTimeout(() => setMessage(''), 3000);
            return;
        }

        try {
            const response = await axios.post(`${baseApiUrl}/digital_medlemsordning/sign_up_activity/`, {
                auth0_id: user.sub,
                activity_id: selectedActivity.activityID,
            });
            if (response.status === 200 || response.status === 201) {
                setMessage('Du er registrert nå.');
                setMessageType('confirmation');
                setTimeout(() => {
                    setMessage('');
                    setSelectedActivity(null);
                    fetchTodayActivities();
                }, 3000);
            }
        } catch (error) {
            console.error('Failed to sign up for the activity:', error);
            setMessage('Du kan ikke registrere nå, vennligst vent litt.');
            setMessageType('error');
            setTimeout(() => setMessage(''), 3000);
        }
    };

    const handleUndoSignUp = async () => {
        if (!isAuthenticated) {
            setMessage('Du må være logget inn for å melde av.');
            setMessageType('error');
            setTimeout(() => setMessage(''), 3000);
            return;
        }

        try {
            const response = await axios.post(`${baseApiUrl}/digital_medlemsordning/undo_signup_activity/`, {
                auth0_id: user.sub,
                activity_id: selectedActivity.activityID,
            });
            if (response.status === 200) {
                setMessage('Du er avmeldt nå.');
                setMessageType('confirmation');
                setTimeout(() => {
                    setMessage('');
                    setSelectedActivity(null);
                    fetchTodayActivities();
                }, 3000);
            }
        } catch (error) {
            console.error('Failed to undo signup for the activity:', error);
            setMessage('Feil ved avmelding, vennligst prøv igjen.');
            setMessageType('error');
            setTimeout(() => setMessage(''), 3000);
        }
    };

    const handleCloseDetails = () => {
        setSelectedActivity(null);
        setMessage('');
    };

    return (
        <div className="today-activities-container">
            <div className="promo-sentence-container">
                <p className={`promo-sentence ${displayedText ? 'visible' : ''}`}>{displayedText}</p>
            </div>
            <div className="today-activities-list">
                {activities.map((activity, index) => (
                    <div key={index} className="today-activity-entry" onClick={() => handleActivityClick(activity)}>
                        <img src={activity.image ? `${baseApiUrl}${activity.image}` : "https://via.placeholder.com/60"} alt={activity.title} className="today-activity-image" />
                        <div className="today-activity-details">
                            <div className="today-activity-title">{activity.title}</div>
                            <div className="today-activity-date">{todayDate ? `Dato: ${todayDate}` : 'Ingen dato oppgitt'}</div>
                        </div>
                    </div>
                ))}
            </div>

            {selectedActivity && (
                <div className="today-details-modal" onClick={handleCloseDetails}>
                    <div className="today-modal-content" onClick={(e) => e.stopPropagation()}>
                        <h2>{selectedActivity.title}</h2>
                        <p>{selectedActivity.description}</p>
                        <div className="today-modal-buttons">
                            <button className="today-modal-button" onClick={handleCloseDetails}>Lukk</button>
                            {selectedActivity.isUserSignedUp ? (
                                <button className="today-modal-button" onClick={handleUndoSignUp}>Meld av</button>
                            ) : (
                                <button className="today-modal-button" onClick={handleSignUp}>Meld på</button>
                            )}
                        </div>
                        {message && (
                            <div className={`message ${messageType === 'confirmation' ? 'confirmation' : 'error'}`}>{message}</div>
                        )}
                    </div>
                </div>
            )}
        </div>
    );
}

export default TodayActivitiesComponent;
