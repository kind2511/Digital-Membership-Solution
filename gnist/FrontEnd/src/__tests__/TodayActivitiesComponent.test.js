import React from 'react';
import renderer from 'react-test-renderer';
import TodayActivitiesComponent from '../components/TodayActivitiesComponent';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { Auth0Context } from '@auth0/auth0-react';

// Mock axios for the test
const mock = new MockAdapter(axios);
mock.onGet('http://127.0.0.1:8000/digital_medlemsordning/get_activity_today/').reply(200, []);

// Mock Auth0 Context
const mockAuth0Context = {
  user: { sub: 'auth0|123456789' },
  isAuthenticated: true,
  loginWithRedirect: jest.fn(),
  logout: jest.fn()
};

test('TodayActivitiesComponent snapshot', () => {
  const component = renderer.create(
    <Auth0Context.Provider value={mockAuth0Context}>
      <TodayActivitiesComponent />
    </Auth0Context.Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
