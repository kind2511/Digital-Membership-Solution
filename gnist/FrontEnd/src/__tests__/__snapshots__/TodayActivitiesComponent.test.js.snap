// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodayActivitiesComponent snapshot 1`] = `
<div
  className="today-activities-container"
>
  <div
    className="promo-sentence-container"
  >
    <p
      className="promo-sentence "
    />
  </div>
  <div
    className="today-activities-list"
  />
</div>
`;
